camel/types/unified_model_type.py

```diff
@@ -0,0 +1,99 @@
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+# Licensed under the Apache License, Version 2.0 (the “License”);
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an “AS IS” BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+
+from threading import Lock
+from typing import TYPE_CHECKING, ClassVar, Dict, Union, cast
+
+if TYPE_CHECKING:
+    from camel.types import ModelType
+
+
+class UnifiedModelType(str):
+    r"""Class used for support both :obj:`ModelType` and :obj:`str` to be used
+    to represent a model type in a unified way. This class is a subclass of
+    :obj:`str` so that it can be used as string seamlessly.
+
+    Args:
+        value (Union[ModelType, str]): The value of the model type.
+    """
+
+    _cache: ClassVar[Dict[str, "UnifiedModelType"]] = {}
+    _lock: ClassVar[Lock] = Lock()
+
+    def __new__(cls, value: Union["ModelType", str]) -> "UnifiedModelType":
+        with cls._lock:
+            if value not in cls._cache:
+                instance = super().__new__(cls, value)
+                cls._cache[value] = cast(UnifiedModelType, instance)
+            else:
+                instance = cls._cache[value]
+        return instance
+
+    def __init__(self, value: Union["ModelType", str]) -> None:
+        pass
+
+    @property
+    def value_for_tiktoken(self) -> str:
+        r"""Returns the model name for TikToken."""
+        return "gpt-4o-mini"
+
+    @property
+    def token_limit(self) -> int:
+        r"""Returns the token limit for the model."""
+        return -1
+
+    @property
+    def is_openai(self) -> bool:
+        r"""Returns whether the model is an OpenAI model."""
+        return True
+
+    @property
+    def is_anthropic(self) -> bool:
+        r"""Returns whether the model is an Anthropic model."""
+        return True
+
+    @property
+    def is_azure_openai(self) -> bool:
+        r"""Returns whether the model is an Azure OpenAI model."""
```

Comment:
The `property` is a bit weird since they are all true.

Reply:
Hey @lightaime , I think here we want to set default value for the property so that if user pass one string value to `model_type`, it will still able to pass the check in ModelFactory

```
        elif model_platform.is_mistral and model_type.is_mistral:
            model_class = MistralModel
```

```
model = ModelFactory.create(
    model_platform=ModelPlatformType.MISTRAL,
    model_type="mistral-large-latest",
    model_config_dict=MistralConfig(temperature=0.0).as_dict(),
)
```

Reply:
I see. But will this lead to a bug if a `string` accidentally get checked  `is_openai` if it is a string for mistral.

Reply:
@WHALEEYE  and me plan to further refactor the ModelFactory part, we will think of a better design for this part

Reply:
### ### 

> I see. But will this lead to a bug if a `string` accidentally get checked `is_openai` if it is a string for mistral.

@lightaime, about this, is there anyway for us to know if a string is for mistral or openai?